### DISCUSSION TITLE
Is there a same object check in Kani?

### DISCUSSION THREAD

**Discussion by QinyuanWu** [model-checking/verify-rust-std#110] [Q&A] [answered]
Is there a way for Kani to check whether two pointers are pointing to the same object, similar to [`__CPROVER_same_object`](https://diffblue.github.io/cbmc/cprover-manual/md_memory_primitives.html)? For example, in [`non_null::offset_from`](https://github.com/model-checking/verify-rust-std/pull/93) one of the desired `requires` clause is that the two pointers passed in should be pointing to the same object. Thanks!
@zhassan-aws @Jimmycreative

**Comment by zhassan-aws** [comment]
A `same_allocation` API was merged in https://github.com/model-checking/kani/pull/3583. It should be available for usage in this repo soon.

**Reply by xsxszab** [reply]
In [Challenge 3](https://github.com/model-checking/verify-rust-std/blob/main/doc/src/challenges/0003-pointer-arithmentic.md), we need to verify pointers to unit types. However, the `same_allocation` api currently doesn't support this when used in function contracts. For example:
```rust
#[ensures(|result| kani::mem::same_allocation(self as *const T, *result as *const T))]
pub const unsafe fn add(self, count: usize) -> Self
```
If `T` is a unit type, the code compiles, but during verification, it produces the following error:
```
Failed Checks: Kani does not support reasoning about pointers to unallocated memory.
```
Is there a workaround to enable `same_allocation` to work with unit types?

**Reply by QinyuanWu** [reply]
@xsxszab Have you tried casting a pointer with valid memory allocation to unit using `as *mut ()`? For example, `trait_object as *const dyn SampleTrait as *mut ()`

**Reply by xsxszab** [reply]
> @xsxszab Have you tried casting a pointer with valid memory allocation to unit using `as *mut ()`? For example, `trait_object as *const dyn SampleTrait as *mut ()`

I've tried it using the following code and it produced the same failed check:
```rust
#[kani::proof]
fn test_unit_type() {
  let unit_obj = ();
  let ptr1: *const () = &unit_obj;
  let ptr2: *const () = &unit_obj;
  assert!(kani::mem::same_allocation(ptr1, ptr2));
}
```

**Reply by zhassan-aws** [reply]
@xsxszab Please file an issue in the [Kani repository](https://github.com/model-checking/kani).

**Reply by xsxszab** [reply]
> @xsxszab Please file an issue in the [Kani repository](https://github.com/model-checking/kani).

Sure, the tracking issue is [here](https://github.com/model-checking/kani/issues/3650#issue-2614994750).

**Reply by celinval** [reply, marked as answer]
> > @xsxszab Have you tried casting a pointer with valid memory allocation to unit using `as *mut ()`? For example, `trait_object as *const dyn SampleTrait as *mut ()`
> 
> I've tried it using the following code and it produced the same failed check:
> 
> ```rust
> #[kani::proof]
> fn test_unit_type() {
>   let unit_obj = ();
>   let ptr1: *const () = &unit_obj;
>   let ptr2: *const () = &unit_obj;
>   assert!(kani::mem::same_allocation(ptr1, ptr2));
> }
> ```

The main problem with this example is that an zero sized type (ZST) object is not allocated. That's why you are getting this error. Your contract has to account to that, i.e.:, check if the type is not a ZST first.
